[SIMD] Intel support for replace_lane and extract_lane

hyjorc1 opened the pull request:
#### c0ff6b0854d6f0e3bcdaf11829d67dea3c5d88e5
<pre>
[SIMD] Intel fix operation shuffle
<a href="https://bugs.webkit.org/show_bug.cgi?id=249073">https://bugs.webkit.org/show_bug.cgi?id=249073</a>
rdar://problem/10321491

Reviewed by NOBODY (OOPS!).

Fix WASM SIMD operation i8x16.shuffle.

* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::addSIMDShuffle):
</pre>

justinmichaud commented on Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
```diff
@@ -505,6 +505,31 @@ class AirIRGenerator {
+        switch (airOp) {
+        case VectorReplaceLaneFloat64:
+            if (isValidForm(airOp, Arg::Imm, Arg::Tmp, Arg::Tmp, Arg::Tmp)) {
+                append(airOp, Arg::imm(imm), s, result, tmpForType(Types::I64));
```
These temps are both GP, so you can make both these cases the same. Also, since you are already checking that the form is valid, you don't need the switch.

justinmichaud commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -4588,6 +4603,86 @@ class X86Assembler {
+            } else if (lane == SIMDLane::i32x4) {
+                ASSERT(laneIndex < 4);
+                // VEX.128.      66.          0F3A.         W0    16 /r ib VPEXTRD r32/m32, xmm2, imm8
+                //     ^isVEX256 ^simdPrefix  ^impliedBytes ^isW1 ^opcode          ^rm      ^reg
```
I'm not sure that this line is helping

justinmichaud commented on Source/JavaScriptCore/assembler/X86Assembler.h:
```diff
@@ -4390,6 +4384,27 @@ class X86Assembler {
                 putByteUnchecked(thirdByte);
             }
+
+            ALWAYS_INLINE void threeBytesVex(bool isVEX256, OneByteOpcodeID simdPrefix, VexImpliedBytes impliedBytes, bool isW1, RegisterID r, RegisterID x, RegisterID b)
```
assert !isVEX256?

hyjorc1 replied:
This function is designed as a universal help method for VEX prefix encoding.

ddegazio commented on Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
```diff
@@ -2140,9 +2140,8 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
     {
-        RegisterID scratch = scratchRegister();
         moveDoubleTo64(src, scratch);
         m_assembler.pinsr(dest, scratch, simdLane, lane.m_value);
```
Do we really need a scratch here? Intel has `(v)insertps` for inserting a scalar float into a vector: https://www.felixcloutier.com/x86/insertps

ddegazio replied:
Seems like it might be a bit more complicated than I'm thinking, we can evaluate this later.

hyjorc1 replied:
Added a FIXME.

justinmichaud reviewed (approved):
r=me

hyjorc1 committed:
[SIMD] Intel support for replace_lane and extract_lane
https://bugs.webkit.org/show_bug.cgi?id=248637

Reviewed by Justin Michaud.

Support WASM SIMD operations replace_lane and extract_lane.
https://github.com/WebAssembly/simd/blob/main/proposals/simd/SIMD.md#accessing-lanes

* Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h:
(JSC::MacroAssemblerX86_64::vectorReplaceLane):
(JSC::MacroAssemblerX86_64::vectorExtractLane):
* Source/JavaScriptCore/assembler/X86Assembler.h:
(JSC::X86Assembler::pinsr):
(JSC::X86Assembler::pextr):
(JSC::X86Assembler::shufps):
(JSC::X86Assembler::shufpd):
(JSC::X86Assembler::X86InstructionFormatter::SingleInstructionBufferWriter::memoryModRM):
* Source/JavaScriptCore/b3/air/AirOpcode.opcodes:
* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::addReplaceLane):

Canonical link: https://commits.webkit.org/257277@main
